File: nnops/cpu/ops/binary_ops.cpp
#include <nnops/cpu/ops/binary_ops.h>
#include <nnops/tensor.h>
#include <nnops/scalar.h>
#include <nnops/data_type.h>
#include <nnops/common.h>

using nnops::Tensor, nnops::TensorMeta, nnops::TensorShape;

namespace nnops::cpu::ops {

struct ScalarBinaryOpParams {
    const scalar_binary_op_t &op;
    const index_t loop;
    index_t offsets[3];
    const index_t strides[3];
};

void do_binary_op_tensor_tensor_impl(const Tensor &self, const Tensor &other, const Tensor &out, int axis, ScalarBinaryOpParams &params) {
    if (axis < self.ndim() - 1) {
        const int loop = self.shape()[axis];
        for (int i = 0; i < loop; i++) {
            do_binary_op_tensor_tensor_impl(self, other, out, axis + 1, params);
            params.offsets[0] += out.stride()[axis];
            params.offsets[1] += self.stride()[axis];
            params.offsets[2] += other.stride()[axis];
        }
        params.offsets[0] -= out.stride()[axis] * loop;
        params.offsets[1] -= self.stride()[axis] * loop;
        params.offsets[2] -= other.stride()[axis] * loop;
        return;
    }

    void *args[3] = {out.data_ptr(params.offsets[0]), self.data_ptr(params.offsets[1]), other.data_ptr(params.offsets[2])};
    params.op(args, params.strides, params.loop);
}

template<ScalarBinaryOpType op_type>
Tensor binary_op_tensor_tensor_template(const Tensor &self, const Tensor &other, bool inplace) {
    Tensor ret, self_br, other_br;
    DataType dtype;

    if (inplace) {
        // TODO: expanded tensor
        NNOPS_CHECK(
            Tensor::is_broadcastable_to(other.shape(), self.shape(), 0),
            "could not broadcast tensor from shape %s into shape %s",
            other.shape_as_string().c_str(), self.shape_as_string().c_str()
        );
        dtype = self.dtype();
        ret = self;
        self_br = self;
        other_br = other.astype(dtype).broadcast_to(self.shape());
    } else {
        NNOPS_CHECK(Tensor::is_broadcastable(self, other),
            "operands could not be broadcast together with shape %s and shape %s",
            self.shape_as_string().c_str(), other.shape_as_string().c_str()
        );
        TensorShape shape = Tensor::broadcast_shape(self, other);
        dtype = get_promote_type(op_type, self.dtype(), other.dtype());
        self_br = self.astype(dtype).broadcast_to(shape);
        other_br = other.astype(dtype).broadcast_to(shape);
        ret = Tensor(dtype, shape, self.device());
    }

    auto scalar_binary_op = get_scalar_binary_op(op_type, dtype);
    ScalarBinaryOpParams params = {
        scalar_binary_op, self_br.shape(-1), {
            0, 0, 0
        }, {
            (index_t)(ret.stride(-1) * ret.itemsize()),
            (index_t)(self_br.stride(-1) * self_br.itemsize()),
            (index_t)(other_br.stride(-1) * other_br.itemsize())
        }
    };

    do_binary_op_tensor_tensor_impl(self_br, other_br, ret, 0, params);

    return ret;
}

#define MAKE_BINARY_OP_TENSOR_TENSOR_FUNCTOR(op_type, op_name, op) \
template Tensor binary_op_tensor_tensor_template<op_type>(const Tensor &self, const Tensor &other, bool inplace);
SCALAR_BINARY_OP_GEN_TEMPLATE_LOOPx1(MAKE_BINARY_OP_TENSOR_TENSOR_FUNCTOR)
#undef MAKE_BINARY_OP_TENSOR_TENSOR_FUNCTOR

template<ScalarBinaryOpType op_type>
Tensor binary_op_tensor_scalar_template(const Tensor &self, const Scalar &other, bool inplace) {
    Tensor other_tensor = other.tensor();
    return binary_op_tensor_tensor_template<op_type>(self, other_tensor, inplace);
}

template<ScalarBinaryOpType op_type>
Tensor binary_op_tensor_scalar_template_reverse(const Scalar &other, const Tensor &self, bool inplace) {
    Tensor other_tensor = other.tensor();
    return binary_op_tensor_tensor_template<op_type>(other_tensor, self, inplace);
}

#define MAKE_BINARY_OP_TENSOR_SCALAR_FUNCTOR(op_type, op_name, op) \
template Tensor binary_op_tensor_scalar_template<op_type>(const Tensor &self, const Scalar &other, bool inplace); \
template Tensor binary_op_tensor_scalar_template_reverse<op_type>(const Scalar &other, const Tensor &self, bool inplace);
SCALAR_BINARY_OP_GEN_TEMPLATE_LOOPx1(MAKE_BINARY_OP_TENSOR_SCALAR_FUNCTOR)
#undef MAKE_BINARY_OP_TENSOR_SCALAR_FUNCTOR

} // namespace nnops::cpu::ops

File: src/cpu/ops/binary_ops.rs
use crate::common::{Index, is_broadcastable, is_broadcastable_to, broadcast_shape};
use crate::data_type::{ScalarBinaryOpType, DataType, get_promote_type, scalar_binary_op};
use crate::tensor::Tensor;
use crate::scalar::Scalar;

struct ScalarBinaryOpParams {
    op_type: ScalarBinaryOpType,
    dtype: DataType,
    loop_size: Index,
    offsets: [Index; 3],
    strides: [Index; 3],
}

fn calc_slice_span(itemsize: Index, stride: Index, size: Index) -> usize {
    if size <= 0 { return 0; }
    let s = stride as usize;
    s * (size as usize - 1) + itemsize as usize
}

fn do_binary_op_tensor_tensor_impl(
    self_t: &Tensor,
    other: &Tensor,
    out: &Tensor,
    axis: usize,
    params: &mut ScalarBinaryOpParams,
) {
    if axis < self_t.ndim() - 1 {
        let loop_size = self_t.shape()[axis];
        for _ in 0..loop_size {
            do_binary_op_tensor_tensor_impl(self_t, other, out, axis + 1, params);
            params.offsets[0] += out.stride_at(axis as Index);
            params.offsets[1] += self_t.stride_at(axis as Index);
            params.offsets[2] += other.stride_at(axis as Index);
        }
        params.offsets[0] -= out.stride_at(axis as Index) * loop_size;
        params.offsets[1] -= self_t.stride_at(axis as Index) * loop_size;
        params.offsets[2] -= other.stride_at(axis as Index) * loop_size;
        return;
    }

    let out_span = calc_slice_span(out.itemsize(), params.strides[0], params.loop_size);
    let self_span = calc_slice_span(self_t.itemsize(), params.strides[1], params.loop_size);
    let other_span = calc_slice_span(other.itemsize(), params.strides[2], params.loop_size);

    let out_slice = unsafe {
        std::slice::from_raw_parts_mut(
            out.data_ptr_with_offset(params.offsets[0]) as *mut u8,
            out_span,
        )
    };
    let self_slice = unsafe {
        std::slice::from_raw_parts(self_t.data_ptr_with_offset(params.offsets[1]), self_span)
    };
    let other_slice = unsafe {
        std::slice::from_raw_parts(other.data_ptr_with_offset(params.offsets[2]), other_span)
    };
    scalar_binary_op(
        params.op_type, params.dtype,
        out_slice, self_slice, other_slice,
        params.strides[0], params.strides[1], params.strides[2],
        params.loop_size,
    );
}

pub fn binary_op_tensor_tensor(op_type: ScalarBinaryOpType, self_t: &Tensor, other: &Tensor, ret: &mut Tensor) {
    let (self_ref, other_ref, dtype);
    let out_ref;

    if std::ptr::eq(self_t, ret) {
        nnops_check!(
            is_broadcastable_to(other.shape(), self_t.shape(), 0),
            "could not broadcast tensor from shape {} into shape {}",
            other.shape_as_string(),
            self_t.shape_as_string()
        );
        dtype = self_t.dtype();
        self_ref = self_t.clone();
        other_ref = other.astype(dtype).broadcast_to_shape(self_t.shape(), 0);
        out_ref = self_t.clone();
    } else {
        nnops_check!(
            is_broadcastable(self_t.shape(), other.shape(), 0),
            "operands could not be broadcast together with shape {} and shape {}",
            self_t.shape_as_string(),
            other.shape_as_string()
        );
        let shape = broadcast_shape(self_t.shape(), other.shape(), 0);
        dtype = get_promote_type(op_type, self_t.dtype(), other.dtype());
        self_ref = self_t.astype(dtype).broadcast_to_shape(&shape, 0);
        other_ref = other.astype(dtype).broadcast_to_shape(&shape, 0);
        out_ref = Tensor::with_device_type(dtype, &shape, self_t.device_type());
    }

    let mut params = ScalarBinaryOpParams {
        op_type,
        dtype,
        loop_size: self_ref.shape_at(-1),
        offsets: [0, 0, 0],
        strides: [
            out_ref.stride_at(-1) * out_ref.itemsize(),
            self_ref.stride_at(-1) * self_ref.itemsize(),
            other_ref.stride_at(-1) * other_ref.itemsize(),
        ],
    };

    do_binary_op_tensor_tensor_impl(&self_ref, &other_ref, &out_ref, 0, &mut params);
    *ret = out_ref;
}

pub fn binary_op_tensor_scalar(op_type: ScalarBinaryOpType, self_t: &Tensor, other: &Scalar, ret: &mut Tensor) {
    let other_tensor = other.to_tensor();
    binary_op_tensor_tensor(op_type, self_t, &other_tensor, ret);
}

pub fn binary_op_tensor_scalar_reverse(op_type: ScalarBinaryOpType, other: &Scalar, self_t: &Tensor, ret: &mut Tensor) {
    let other_tensor = other.to_tensor();
    binary_op_tensor_tensor(op_type, &other_tensor, self_t, ret);
}

pub fn add(self_t: &Tensor, other: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_tensor(ScalarBinaryOpType::Add, self_t, other, &mut ret);
    ret
}

pub fn sub(self_t: &Tensor, other: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_tensor(ScalarBinaryOpType::Sub, self_t, other, &mut ret);
    ret
}

pub fn mul(self_t: &Tensor, other: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_tensor(ScalarBinaryOpType::Mul, self_t, other, &mut ret);
    ret
}

pub fn truediv(self_t: &Tensor, other: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_tensor(ScalarBinaryOpType::Div, self_t, other, &mut ret);
    ret
}

pub fn add_scalar(self_t: &Tensor, other: &Scalar) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar(ScalarBinaryOpType::Add, self_t, other, &mut ret);
    ret
}

pub fn sub_scalar(self_t: &Tensor, other: &Scalar) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar(ScalarBinaryOpType::Sub, self_t, other, &mut ret);
    ret
}

pub fn mul_scalar(self_t: &Tensor, other: &Scalar) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar(ScalarBinaryOpType::Mul, self_t, other, &mut ret);
    ret
}

pub fn truediv_scalar(self_t: &Tensor, other: &Scalar) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar(ScalarBinaryOpType::Div, self_t, other, &mut ret);
    ret
}

pub fn add_scalar_reverse(other: &Scalar, self_t: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar_reverse(ScalarBinaryOpType::Add, other, self_t, &mut ret);
    ret
}

pub fn sub_scalar_reverse(other: &Scalar, self_t: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar_reverse(ScalarBinaryOpType::Sub, other, self_t, &mut ret);
    ret
}

pub fn mul_scalar_reverse(other: &Scalar, self_t: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar_reverse(ScalarBinaryOpType::Mul, other, self_t, &mut ret);
    ret
}

pub fn truediv_scalar_reverse(other: &Scalar, self_t: &Tensor) -> Tensor {
    let mut ret = Tensor::new();
    binary_op_tensor_scalar_reverse(ScalarBinaryOpType::Div, other, self_t, &mut ret);
    ret
}